### PR TITLE
fix: analytics timeout

### PR DESCRIPTION
**Pull request by mstasgravitee**
fixes AM-1110

two things: 
on the backend added extended index with outcome.status == 'SUCCESS'  
on the frontend made dashboard resilient to single indicator crash. 

for count pipeline aggregate with old index () the 3 month request for 0.5M user logins took `executionTimeMillis: 1150`
with new index `executionTimeMillis: 154`
the proof the planner takes new index: 

> winningPlan: {
            stage: 'COUNT_SCAN',
            keyPattern: {
              referenceType: 1,
              referenceId: 1,
              type: 1,
              'outcome.status': 1,
              timestamp: -1
            },
            indexName: 'r1ty1s1t_1'  <-- NEW ONE

Mockups: 
https://www.figma.com/file/k7okoie28WAjDjsnnEAjFc/AM---Dashboard?type=design&node-id=119-3190&mode=design&t=IPYa2HjD9JFmxrc0-0

Dashboard after changes:
image

image

image

image

### CODE REVIEW

File: gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java [hunk 1]
```diff
@@ -115,7 +116,7 @@ public class MongoAuditReporter extends AbstractService<Reporter> implements Aud
     @Value("${management.mongodb.ensureIndexOnStart:true}")
 
     private boolean ensureIndexOnStart;
-    @Value("${management.mongodb.cursorMaxTime:60000}")
+    @Value("${management.mongodb.cursorMaxTime:28000}")
```
Comment (leleueri) [inline gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java:119]: question: why did you change this default cursorMaxTime value for this one ?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java [hunk 2]
```diff
@@ -378,7 +377,7 @@ public String createReporterConfig(String domain) {
 
     private String addOptionsToURI(Environment environment, String mongoUri) {
         Integer connectTimeout = environment.getProperty("management.mongodb.connectTimeout", Integer.class, 1000);
-        Integer socketTimeout = environment.getProperty("management.mongodb.socketTimeout", Integer.class, 1000);
+        Integer socketTimeout = environment.getProperty("management.mongodb.socketTimeout", Integer.class, 29000);
```
Comment (leleueri) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java:380]: **question**: why 29s and not 30 ?

Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java:380]: The 30s is the HTTP timeout and socket timeout has to be set lower than 30s to see this error instead of HTTP timeout.

File: gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java (hunk 1 above)
Comment (remisultan) [inline gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java:119]: Why the value at `28000` ?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java [hunk 3]
```diff
@@ -347,7 +346,7 @@ public String createReporterConfig(String domain) {
             // dash are forbidden in table name, replace them in domainName by underscore
             String tableSuffix = null;
             if (domain != null) {
-                tableSuffix = domain.replaceAll("-", "_");
+                tableSuffix = domain.replace("-", "_");
```
Comment (remisultan) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java:349]: As per the comment 
 ```
   // dash are forbidden in table name, replace them in domainName by underscore
```

Why use `replace` instead of `replaceAll` ?

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java (hunk 2 above)
Comment (remisultan) [inline gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java:380]: Why is the value at `29000` ?

**Review by remisultan — COMMENTED** [review]
A few comments.

Notably on the choice of default values and the `replaceAll` change

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java (hunk 3 above)
Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java:349]: replaceAll uses regex as input

replace uses CharSequence as input

File: gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java (hunk 1 above)
Comment (mstasgravitee) [reply gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java:119]: The curson max time has to be lower than socket timeout

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java (hunk 2 above)
Comment (mstasgravitee) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java:380]: The value has to be lower than HTTP timeout (30s)

File: gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java (hunk 3 above)
Comment (remisultan) [reply gravitee-am-service/src/main/java/io/gravitee/am/service/impl/ReporterServiceImpl.java:349]: Thanks. You reminded me how it was confusing the naming of these methods were 😄 , `replace` does indeed replace all.

File: gravitee-am-ui/src/app/utils/time-range-utils.ts [hunk 4]
```diff
@@ -45,29 +45,29 @@ export const availableTimeRanges: TimeRange[] = [
...
   },
   {
     id: '30d',
     name: 'This month',
-    value: 1,
-    unit: 'months',
-    interval: 1000 * 60 * 60 * 24,
+    value: 4,
+    unit: 'weeks',
+    interval: 1000 * 60 * 60 * 24 * 7,
```
Comment (leleueri) [inline gravitee-am-ui/src/app/utils/time-range-utils.ts:64]: **thought**: I'm wondering if we will have enough interval to make the graph useful :thinking:

File: gravitee-am-ui/src/app/utils/time-range-utils.ts [hunk 5]
```diff
@@ -45,29 +45,29 @@ export const availableTimeRanges: TimeRange[] = [
...
   },
   {
     id: '90d',
     name: 'Last 90 days',
     value: 3,
     unit: 'months',
-    interval: 1000 * 60 * 60 * 24,
+    interval: 1000 * 60 * 60 * 24 * 30,
```
Comment (leleueri) [inline gravitee-am-ui/src/app/utils/time-range-utils.ts:71]: **thought**: I'm wondering if we will have enough interval to make the graph useful 🤔

Comment (remisultan) [reply gravitee-am-ui/src/app/utils/time-range-utils.ts:71]: Thought the same. It may not return enough info for the graph to be useful

File: gravitee-am-ui/src/app/utils/time-range-utils.ts (hunk 4 above)
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/utils/time-range-utils.ts:64]: what do you propose?

File: gravitee-am-ui/src/app/utils/time-range-utils.ts (hunk 5 above)
Comment (mstasgravitee) [reply gravitee-am-ui/src/app/utils/time-range-utils.ts:71]: what do you propose?

File: gravitee-am-ui/src/app/utils/time-range-utils.ts (hunk 4 above)
Comment (leleueri) [reply gravitee-am-ui/src/app/utils/time-range-utils.ts:64]: For a Month, I think a one day interval is good.

File: gravitee-am-ui/src/app/utils/time-range-utils.ts (hunk 5 above)
Comment (leleueri) [reply gravitee-am-ui/src/app/utils/time-range-utils.ts:71]: For a 3 month period, maybe a one week interval could be enough but a one month interval is not enough IMO. You should ask to product team.

File: gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/constants/MongoAuditReporterConstants.java [hunk 6]
```diff
@@ -38,6 +38,7 @@ private MongoAuditReporterConstants(){}
   public static final String FIELD_ACCESS_POINT_ID = "accessPoint.id";
   public static final String INDEX_REFERENCE_TIMESTAMP_NAME = "r1t_1";
   public static final String INDEX_REFERENCE_TYPE_TIMESTAMP_NAME = "r1ty1t_1";
+  public static final String INDEX_REFERENCE_TYPE_STATUS_SUCCESS_TIMESTAMP_NAME = "r1ty1success1t_1";
```
Comment (remisultan) [inline gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/constants/MongoAuditReporterConstants.java:41]: Is the index name short enough ?
Thinking of DocumentDB

Comment (lgw-gravitee) [reply gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/constants/MongoAuditReporterConstants.java:41]: I shorted it to r1ty1s1t_1
same length like existing INDEX_REFERENCE_ACTOR_ID_TARGET_ID_TIMESTAMP_NAME (10)

File: gravitee-am-ui/src/app/components/widget/widget.component.html [hunk 7]
```diff
@@ -20,7 +20,14 @@
     <mat-card-title>{{ widget.title }}</mat-card-title>
     <mat-card-subtitle>{{ widget.subhead }}</mat-card-subtitle>
   </mat-card-header>
-  <mat-card-content>
+  <mat-card-content class="no-data-card" *ngIf="!widget.chart.response"
```
Comment (mstasgravitee) [inline gravitee-am-ui/src/app/components/widget/widget.component.html:23]: please use lint:fix

File: gravitee-am-ui/src/app/components/widget/widget.component.html [hunk 8]
```diff
@@ -20,7 +20,14 @@
...
+                    matTooltip="The data for this category could not be fetched. This could be related to the amount of data to process. Try to change the Time Range filter to see results.">
+    <div class="no-data-text-block">
+      <mat-icon style="margin-right: 5px">info_outline</mat-icon>
+      <span class="no-data">Failed to fetch data</span>
```
Comment (mstasgravitee) [inline gravitee-am-ui/src/app/components/widget/widget.component.html:27]: I don't see "no-data" class

File: gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java [hunk 9]
```diff
@@ -240,6 +242,7 @@ private void initIndexes() {
             Map<Document, IndexOptions> indexes = new HashMap<>();
             indexes.put(new Document(FIELD_REFERENCE_TYPE, 1).append(FIELD_REFERENCE_ID, 1).append(FIELD_TIMESTAMP, -1), new IndexOptions().name(INDEX_REFERENCE_TIMESTAMP_NAME));
             indexes.put(new Document(FIELD_REFERENCE_TYPE, 1).append(FIELD_REFERENCE_ID, 1).append(FIELD_TYPE, 1).append(FIELD_TIMESTAMP, -1), new IndexOptions().name(INDEX_REFERENCE_TYPE_TIMESTAMP_NAME));
+            indexes.put(new Document(FIELD_REFERENCE_TYPE, 1).append(FIELD_REFERENCE_ID, 1).append(FIELD_TYPE, 1).append(FIELD_STATUS, 1).append(FIELD_TIMESTAMP, -1), new IndexOptions().name(INDEX_REFERENCE_TYPE_STATUS_SUCCESS_TIMESTAMP_NAME).partialFilterExpression(new Document(FIELD_STATUS, new Document("$eq", "SUCCESS"))));
```
Comment (mstasgravitee) [inline gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java:245]: question: why is partialFilterExpression for SUCCESS? What is about  FAILURE which is also used in queries?

Comment (tcompiegne) [inline gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java:245]: Do we have some indexes to set for the JDBC (SQL) implementation ?

**Review by tcompiegne — CHANGES_REQUESTED** [review]

File: gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java (hunk 9 above)
Comment (lgw-gravitee) [reply gravitee-am-reporter/gravitee-am-reporter-mongodb/src/main/java/io/gravitee/am/reporter/mongodb/audit/MongoAuditReporter.java:245]: Dont think we can provide such. The query for JDBC requires more in depth analysis as it has a few JOINs. 
Even though I added index, it didn't make a query faster nor didn't remove SeqScans.

image
 
I would leave JDBCs indexes intact

**Review by remisultan — APPROVED** [review]
LGTM
